test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py

```diff
@@ -0,0 +1,428 @@
+                                },
+                                "vrf_public_key": new_vrf_public_key,
+                                "decommission_key": "AnyoneCanSpend",
+                                "margin_ratio_per_thousand": 1000,
+                                "cost_per_block" : "0"
+                            },
+                        ],
+                    },
+                ],
+            },
+            "signatures": [
+                {
+                    "NoSignature": None,
+                },
+            ],
+        }
+
+        (create_new_pool_transaction_encoded, create_new_pool_transaction_id) = self.pack_transaction(create_new_pool_transaction)
+        self.generate_block(2, block_input_data, [create_new_pool_transaction_encoded])
+
+        #
+        # Stake many blocks with the new stake pool
+        #
+
+        kernel_input_outpoint = {
+            "id": {
+                "Transaction": self.hex_to_dec_array(create_new_pool_transaction_id),
+            },
+            "index": 0,
+        }
+
+        kernel_input_utxo = {
+            "CreateStakePool": [
+                new_pool_id,
+                {
+                    "value": "10000000000000000",
+                    "staker": {
+                        "PublicKey": new_stake_public_key,
+                    },
+                    "vrf_public_key": new_vrf_public_key,
+                    "decommission_key": "AnyoneCanSpend",
+                    "margin_ratio_per_thousand": 1000,
+                    "cost_per_block" : "0"
+                },
+            ],
+        }
+
+        for i in range(0, random.randint(10,100)):
+            block_input_data = block_input_data_obj.encode({
+                "PoS": {
+                "stake_private_key": new_stake_private_key,
+                    "vrf_private_key": new_vrf_private_key,
+                    "pool_id": new_pool_id,
+                    "kernel_inputs": [
+                        {
+                            "Utxo": kernel_input_outpoint,
+                        },
+                    ],
+                    "kernel_input_utxo": [kernel_input_utxo]
+                }
+            }).to_hex()[2:]
+
+            self.generate_block(3 + i, block_input_data, [])
+
+            kernel_input_outpoint = {
+                "id": {
+                    "BlockReward": self.hex_to_dec_array(self.nodes[0].chainstate_best_block_id()),
+                },
+                "index": 0,
+            }
+
+            kernel_input_utxo = {
+                "ProduceBlockFromStake": [
+                    {
+                        "PublicKey": new_stake_public_key,
+                    },
+                    new_pool_id,
+                ],
+            }
+
+if __name__ == '__main__':
+    # Use the first instance to bootstrap new staking keys
+    GenerateGenesisKeys().main(exit_on_success=False)
+    GeneratePoSBlocksTest().main()
```

Comment:
Could you please explain what's going on here? Looks like a hack :laughing:

Reply:
Hack, magic... it's all perspective 😂